### PR TITLE
Schedule Reports

### LINKED ISSUE #2779: Export setup - report schedule
_This ticket is part of epic #2626 and issue #2658_

### Detailed description
After choosing reports type, selecting objects and configuring plugins the user will come to the export setup page. In general the page consists of the same layout as other pages in the flow: XXL header 'Generate a report', a small description and the stepper that is filled to 'Export setup'. Following is a button that offers the user to go back to change the selection, and the start of the page 'Export setup' with a small description text.

To end the content body down the page there is a primary button to 'Generate report' and a  illustration of a line and a fish (like in the stepper) meaning that this is the end of the generate report flow.

### Specifications
The implementation should include…
- [x] Page title and stepper
- [x] Section title and description
- [x] Date picker
- [x] Recurrence dropdown (based on the date picker)
- [ ] Option in dropdown to create a custom recurrence (opens a modal)
- [x] 'Generate report' button
- [x] Activation of the scheduler in the backend
- [x] Footer

### Screenshots
afbeelding
Image

### Figma link
Updated implementation Figma file link: https://www.figma.com/design/5lYYfE1SqJ9qofjyvwdvtw/%E2%9C%85-KAT-%7C-Approved-%7C-Reports-Flow?node-id=2325%3A13868&t=yzPYrYFsQAygovBO-1

### Possible solution
_Outline your proposed solution for implementing the feature. You can include any specific ideas, designs, or functionalities here._

### Alternatives considered
_Describe any alternative approaches or solutions you've considered, and why you believe the proposed solution is superior._

### LINKED ISSUE #3088: Create JSON for making reports
_This ticket is part of epic #2626_

### Detailed description
This is step 1 in creating scheduled reports. While generating a report, we need to create a JSON.
With every step in the report flow, this JSON needs to be expanded (with a post-request). 
At the end of the flow, the report will be generated based on this JSON. 

This JSON will help us to go back and forth through the generate-report pages.
This will also make it possible to implement other features, like scheduling reports and creating "recipes".

This ticket is pre-work that has to be done before we can implement the scheduled reports.

### Possible solution

The JSON as here described will be "all things needed to make a report". 
This consists of a JSON object that will be passed from view to view with user selections.

What is needed for this JSON?
**- OOIs:** Remember which objects are selected. We don't want to make multiple octopoes requests each time we need the show attributes. 
**- OOIs filters.** We want to remember the filters used for filtering OOIs
- observed_at: When are these OOIs observed when creating the report.
**- Report types:** Which report types are selected?
**- Plugins:** We want to remember the state of the plugins during report creation.

With this information we can generate data for the report types selected: the data together with the JSON we will save in bytes.
The data consists of:
- Calculations to find/fetch vulnerabilities.
- Aggregation of data for aggregate report (the sum of parts) (de-duplication)
- Subreports 

This is an example of how the JSON should look like. If also configs are needed, this can be added to the JSON.
```
report = {
   "observed_at" : "2014-06-13 23:59:59 00:00",
   "ooi_filters": 
      {
           "ooi_types": ["Hostname","URL"]
           "clearance_levels": [2, 3, 4],
           "clearance_types": "["declared", "inherited"]
      },
   "oois" : 
       {
           "pk": "Hostname|internet|mispo.es",
           "clearance_level": 2,
           "clearance_type": "declared"
        },
        {
           "pk": "Hostname|internet|mispo.es",
           "clearance_level": 2,
           "clearance_type": "declared"
        },
    "report_types": ["aggregate_report", dns_report", "findings_report"],
    "plugins": {
        "plugin_id": "dns-records",
        "name": "DnsRecords",
        "description": "Fetch the DNS record(s) of a hostname",
        "type": "boefje",
        "options": "required",
        "scan_level": 2,
      },
       "plugin_id": "dns-records",
        "name": "DnsRecords",
        "description": "Fetch the DNS record(s) of a hostname",
        "type": "boefje",
        "options": "required",
        "scan_level": 2,
        
}
```
### Where do we save report recipes? 
There are some options:
- [x] Create a new Report Recipe OOI and/or add a new OOi Type
- [ ] Report OOI. Save the recipe in a variable so it can be used to generate another report in time. When the OOI is rescheduled we can use the recipe info send to Scheduler to unpack it there. If a report task means, do a rerun for each ooi in the recipe. I am doubting about the plugin information needs to be stored in the recipe. 
image

- [ ] Scheduler. We can also feed the Scheduler the Report OOI with the recipe in it and then extract the data in Scheduler.

- [ ] Report Data. Report Data is stored in bytes. 
image

use also in:
image

To unpack the data for the view.

### PR DESCRIPTION
**Pull request by Rieven**
### Changes

- Added form to schedule reports
- Added `backend` functionality to schedule reports in the scheduler @jpbruinsslot 
- Show form at final settings just before you generate report

TO DO in base branch `feature/report-recipe`
- Create report recipe #3442 
- Post recipe with schedule and cron to the scheduler

Depends on:
- Task runner for schedule reports @Donnype

Upcoming features:
- Show report schedules in a list
- Change schedules

### Issue link
#3088 
#2779 

Closes #3088 
Closes #2779 

### Demo
https://github.com/user-attachments/assets/d775734f-dcf0-494f-9f58-3d19abffbcb7

### QA notes

See video for testing.

### Code Checklist

- [x] All the commits in this PR are properly PGP-signed and verified.
- [x] This PR only contains functionality relevant to the issue.
- [ ] I have written unit tests for the changes or fixes I made.
- [x] I have checked the documentation and made changes where necessary.
- [x] I have performed a self-review of my code and refactored it to the best of my abilities.

- [x] Tickets have been created for newly discovered issues.
- [x] For any non-trivial functionality, I have added integration and/or end-to-end tests.
- [x] I have informed others of any required `.env` changes files if required and changed the `.env-dist` accordingly.
- [x] I have included comments in the code to elaborate on what is not self-evident from the code itself, including references to issues and discussions online, or implicit behavior of an interface.

---

## Checklist for code reviewers:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---

## Checklist for QA:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.

### CODE REVIEW

File: rocky/rocky/views/scheduler.py [hunk 1]
```diff
@@ -61,12 +75,49 @@ def get_task_list(self) -> LazyTaskList | list[Any]:
...
+
     def get_task_details(self, task_id: str) -> Task | None:
         try:
             return self.scheduler_client.get_task_details(task_id)
         except SchedulerError as error:
             return messages.error(self.request, error.message)
 
+    def create_report_schedule(self, report_ooi, start_date: str, recurrence: str) -> ScheduleResponse | None:
+        try:
+            self.bytes_client.get_raw(report_ooi.data_raw_id)
```
Comment (Rieven) [inline rocky/rocky/views/scheduler.py:93]: bytes won't be necessary anymore once we crate the recipe. The recipe will recreate the data and add it to bytes.

File: rocky/reports/forms.py [hunk 2]
```diff
@@ -28,3 +30,97 @@ def __init__(self, report_types: set[Report], *args, **kwargs):
...
+            if day == 29:
+                self.add_error("start_date", _("Warning: Recurrence is set in February only for leap years."))
+                return False
+            if day == 30:
+                self.add_error("start_date", _("Warning: Recurrence is not set for February."))
+                return False
+            if day == 31:
+                self.add_error("start_date", _("Warning: Recurrence will skip months that does not have 31 days."))
+                return False
+        return True
```
Comment (jpbruinsslot) [inline rocky/reports/forms.py:70]: You could perhaps solve this problem by setting the cron expression to `0 0 28-31 * *`  when the start date is in the  28-31 range

**Comment by jpbruinsslot** [comment]
@Rieven do you need some help with resolving the merger? Seems both branches are somewhat out of sync. What is specifically the ordering, this branch into `feature/report-recipe` or does this branch need to be merge into main?

**Comment by Rieven** [comment]
> @Rieven do you need some help with resolving the merger? Seems both branches are somewhat out of sync. What is specifically the ordering, this branch into `feature/report-recipe` or does this branch need to be merged into main?

first merge with `feature/report-recipe` and then everything to main.

**Pull request merged**